doc/programming/parsing.rst

```diff
@@ -921,6 +921,19 @@ properties:
     synchronization point during :ref:`error recovery <error_recovery>`.
     The literal is left in the input stream.
 
+``%synchronize-at = ID;``
+    Specifies a function to synchronize the input. The function must take a
+    ``view<stream>`` as only argument, and return a ``tuple<int8, uint64>``
+    with the following semantics:
+
+    - ``int8`` value is ``1``: synchronization point *is* in the view at the
+      offset specified by the ``uint64`` value.
+    - ``int8`` value is ``0``: synchronization point *could be* in the view,
+      need at least the number of bytes specified in ``uint64`` value to
+      decide.
+    - ``int8`` value is ``-1``: synchronization point not in view and never
+      will, call function again with new data.
+
```

Comment:
Couple thoughts on the API description:

- Suggest to add to the docs that case 0 means buffering: we'll hold on to the current input data until the synchronization functions declares success (1) or failure (0).

- A function could return -1 but still retain state about what it saw, right? So a way to avoid the buffering with case 0 would be a state machine that remembers partial matches, returns -1, and then continues matching with the next data chunk. Might be worth pointing that out in the docs: -1 isn't necessarily a total failure, it's just that the function won't need the data anymore.

Reply:
>  So a way to avoid the buffering with case 0 would be a state machine that remembers partial matches, returns -1, and then continues matching with the next data chunk.

Well, I guess that's assuming that the actual synchronization point follows *after* whatever the function is looking for. If it's at the beginning, we need to buffer.

Reply:
This has the semantics identical to `%synchronize-at` with expression, the function needs to point at data which parses successfully. This means that we have to buffer for `0`, and `1`, and can trim input only for `-1`.